WordPress/Classes/ApplicationToken/ApplicationTokenListView.swift

```diff
@@ -0,0 +1,149 @@
+
+    fileprivate init(tokens: [ApplicationTokenItem]) {
+        let dataProvider = StaticTokenProvider(tokens: .success(tokens))
+        self.init(viewModel: ApplicationTokenListViewModel(dataProvider: dataProvider))
+    }
+
+    fileprivate init(error: Error) {
+        let dataProvider = StaticTokenProvider(tokens: .failure(error))
+        self.init(viewModel: ApplicationTokenListViewModel(dataProvider: dataProvider))
+    }
+
+    init(viewModel: ApplicationTokenListViewModel) {
+        self.viewModel = viewModel
+    }
+
+    var body: some View {
+        VStack {
+            if viewModel.isLoadingData {
+                ProgressView()
+            } else if let error = viewModel.errorMessage {
+                EmptyStateView(Self.errorTitle, systemImage: "exclamationmark.triangle", description: error)
+            } else {
+                List(viewModel.applicationTokens) { token in
+                    ApplicationTokenListItemView(item: token)
+                }
+                .listStyle(.plain)
+            }
+        }
+        .navigationTitle(Self.title)
+        .navigationBarTitleDisplayMode(.inline)
+        .onAppear {
+            if viewModel.applicationTokens.isEmpty {
+                Task {
+                    await viewModel.fetchTokens()
+                }
+            }
+        }
+    }
+}
+
+class ApplicationTokenListViewModel: ObservableObject {
+
+    @Published
+    var isLoadingData: Bool = false
+
+    @Published
+    var errorMessage: String? = nil
+
+    @Published
+    var applicationTokens: [ApplicationTokenItem]
+
+    private let dataProvider: ApplicationTokenListDataProvider!
+
+    init(dataProvider: ApplicationTokenListDataProvider) {
+        self.dataProvider = dataProvider
+        self.applicationTokens = []
+    }
+
+    @MainActor
+    func fetchTokens() async {
+        isLoadingData = true
+        defer {
+            isLoadingData = false
+        }
+
+        do {
+            let tokens = try await self.dataProvider.loadApplicationTokens()
+                .sorted { lhs, rhs in
+                    // The most recently used/created is placed at the top.
+                    switch (lhs.lastUsed, rhs.lastUsed) {
+                    case let (.some(lhsLastUsed), .some(rhsLastUsed)):
+                        return lhsLastUsed > rhsLastUsed
+                    case (.some, .none):
+                        return true
+                    case (.none, .some):
+                        return false
+                    default:
+                        return lhs.createdAt > rhs.createdAt
+                    }
+                }
+            self.applicationTokens = tokens
+        } catch let error as WpApiError {
+            self.errorMessage = error.errorMessage
+        } catch {
+            self.errorMessage = NSLocalizedString("Something went wrong", comment: "A generic error message")
+        }
+    }
+}
+
+private extension WpApiError {
```

Comment:
Assuming that this `WpApiError` will be used throughout the app, it needs to be `LocalizableError` so that the app could simply display it as is.

Reply:
`WpApiError` is generated by uniffi-rs, which comes with a problematic `LocalizedError` implementation. We'll probably roll our own solution in [wordpress-rs's localization work](https://github.com/wordpress-mobile/WordPress-iOS/pull/23484#discussion_r1708418497), which is also part of the reason why I marked this extension as private.